test/test_nn.py

```diff
@@ -3561,8 +3561,134 @@ def mseloss_no_reduce_module_test():
         reference_fn=lambda i, m: (i - target).pow(2))
 
 
+class TestNLLLoss(torch.nn.modules.module.Module):
```

Comment:
Why do we have to add these test modules, plus extra functions that return dicts, which seem to be valid literals that would be specified in the list itself? I don't understand what's the value here

Reply:
I made a module that wraps NLLLoss so that its forward function takes one argument, the input. The [module tests](https://github.com/pytorch/pytorch/blob/d5604aea0bf23c47a915f470c903beb3b6915090/test/common_nn.py#L646) assume that the module's forward function takes one argument and I wanted to reuse the module tests because the criterion tests assume the output of the loss is scalar.

Good point about the return dicts. I think that once I add in reference functions for each of these, I'll need the return dict because a target should be passed to the test module constructor and to the reference function, just like [this](https://github.com/pytorch/pytorch/blob/master/test/test_nn.py#L3639)

Alternatively to make testing cleaner I could write a new sort of criterion test logic that doesn't assume the output of the loss is scalar.

Reply:
You could always use the newly added `wrap_functional` to test a partially applied `NLLLoss` with different arguments, including the targets.

Reply:
Thanks! That was useful